File: решения на phyton/olimp196.py
n = int(input())
arr = [0] * n 
for i in range(n): 
    arr[i] = [0] * n
x=0
y=0
w=1
arr[y][x]=1
a=1
flag=True
while flag==True:
    if w==1:
        while x+1<=n-1 and arr[y][x+1]==0:
            a+=1
            x+=1
            arr[y][x]=a
        w+=1
        if a==n*n:
            flag=False
    elif w==2:
        while y+1<=n-1 and arr[y+1][x]==0:
            a+=1
            y+=1
            arr[y][x]=a
        w+=1
        if a==n*n:
            flag=False
    elif w==3:
        while x-1>=0 and arr[y][x-1]==0:
            a+=1
            x-=1
            arr[y][x]=a
        w+=1
        if a==n*n:
            flag=False
    elif w==4:
        while y-1>=0 and arr[y-1][x]==0:
            a+=1
            y-=1
            arr[y][x]=a
        w+=1
        if a==n*n:
            flag=False
    else:
        w=1
h=len(str(a))
s=''
for y in range(0,n):
    for x in range(0,n):
        arr[y][x]=str(arr[y][x])
        while len(str(arr[y][x]))<h+1:
            arr[y][x]=' '+str(arr[y][x])
        s=s+arr[y][x]
    s=s+'\n'
print(s[:(len(s)-1)])



















        

File: решения на c++/olimp196.cpp
#include <iostream>
int main()
{
    int arr[100][100] = { 0 }, n = 1, c, a = 1;
    unsigned short int x, y;
    bool flag = true;
    for (int py = 0; py < 100; py++) {
        for (int px = 0; px < 100; px++) {
            arr[py][px] = 1;
        }
    }
    std::cin >> c;
    for (int y = 0; y < c; y++) {
        for (int x = 0; x < c; x++) {
            arr[y][x] = 0;
        }
    }
    x = 0;
    y = 0;
    arr[0][0] = 1;
    while (a < (c * c)) {
        if (n == 1) {
            flag = true;
            while (flag == true) {
                a++;
                x++;
                arr[y][x] = a;
                if (x == 99) {
                    flag = false;
                }
                else if (arr[y][x + 1] != 0) {
                    flag = false;
                }
            }
            n++;
        }
        else if (n == 2) {
            flag = true;
            while (flag == true) {
                a++;
                y++;
                arr[y][x] = a;
                if (y == 99) {
                    flag = false;
                }
                else if (arr[y + 1][x] != 0) {
                    flag = false;
                }
            }
            n++;
        }
        else if (n == 3) {
            flag = true;
            while (flag == true) {
                a++;
                x--;
                arr[y][x] = a;
                if (x == 0) {
                    flag = false;
                }
                else if (arr[y][x - 1] != 0) {
                    flag = false;
                }
            }
            n++;
        }
        else if (n == 4) {
            flag = true;
            while (flag == true) {
                a++;
                y--;
                arr[y][x] = a;
                if (y == 0) {
                    flag = false;
                }
                else if (arr[y - 1][x] != 0) {
                    flag = false;
                }
            }
            n++;
        }
        else {
            n = 1;
        }
    }
    for (int py = 0; py < c; py++) {
        for (int px = 0; px < c; px++) {
            std::cout << arr[py][px] << " ";
        }
        std::cout << "\n";
    }
}